### PR TITLE
Credits auto rewind on vote recreation

### LINKED ISSUE #22512: Deleting and recreating vote account causes `calculate_stake_points_and_credits` to exit early
#### Problem
If an active vote account with any previous epoch's of voting history is emptied of lamports and then recreated at the same address, the vote account and delegated stake accounts will stop seeing staking rewards generated due to this case in the stake program evaluating as true: https://github.com/solana-labs/solana/blob/9c9f2dd5bd5d7aee79719fa84797e5fa63db676c/programs/stake/src/stake_state.rs#L218
This is because the recreated vote account starts from scratch with 0 credits, while the stake account remembers all the credits from the original account. The `calculate_stake_points_and_credits` early exit will continue until the observed credits in the recreated vote account surpasses the account's original observed credits, or until the stake account resets its observed credits by deactivating and redelegating to the vote account.

#### Proposed Solution
When this case is first hit (on the first epoch boundary after vote-account recreation), reset the stake-account's observed credits to 0 (and activation epoch to current epoch?) so that next epoch, the stake account starts seeing new credits.

### PR DESCRIPTION
**Pull request by ryoqun**
~~still draft; exploring best-desired least-intrusive code changes with some preparatory code cleanings (should be separated into its own pr)...~~

todo:
- [x] should reset activation_epoch? (i guess we should avoid that and write down the reasons...)
- [x] what about deactivation epoch?
- [x] test, proper feature gate (currently, brokenly piggy-bucking on an existing one)

#### Problem

(See #22512)

#### Summary of Changes

(See #22512)

Fixes #22512

### CODE REVIEW

File: programs/stake/src/stake_state.rs
Comment (ryoqun) [inline programs/stake/src/stake_state.rs:233]: namely, don't force to update for `credits_in_vote == credits_in_stake` both before and after feature activation for compatibility. (add comment).

File: programs/stake/src/stake_state.rs
Comment (ryoqun) [inline programs/stake/src/stake_state.rs:226]: readability: `enum NoCredit:{Delinquent, Recreated}` or `should_rewrite`

Comment (ryoqun) [reply programs/stake/src/stake_state.rs:226]: or `out_of_sync`/`outdated`?

File: programs/stake/src/stake_state.rs
Comment (ryoqun) [reply programs/stake/src/stake_state.rs:233]: maybe, add `// add else if credits_in_vote == credits_in_stake` for later clear guidance for post-feature-activation code cleaning.

**Comment by ryoqun** [comment]
I ran `solana-ledger-tool capitalization --warp-epoch +1` against mainnet-beta. and it seems I'm getting correct inflation tracer outputs:

```
## before

5mHWstuuRk7DVPPVTZNYM43Se9czqBeZ6Woo7QrLi9CD (Stake11111111111111111111111111111111111111): ◎ 432.872655746 => ◎ 432.872655746 (+◎ 0.000000000 0.000000000%) Some(CalculationDetail { epochs: 0, voter: 9Gko8QZBbV5SrEvHKtQHcMrGGSfgFP3KJUozEGifu25x, voter_owner: Vote111111111111111111111111111111111111111, current_effective_stake: 432868089986, total_stake: 432868089986, rent_exempt_reserve: 2282880, points: [], base_rewards: 0, commission: 10, vote_rewards: 0, stake_rewards: 0, activation_epoch: 260, deactivation_epoch: None, point_value: None, old_credits_observed: Some(4169959), new_credits_observed: None, skipped_reasons: "ZeroCreditsAndReturnCurrent/ZeroPoints" })

## after

5mHWstuuRk7DVPPVTZNYM43Se9czqBeZ6Woo7QrLi9CD (Stake11111111111111111111111111111111111111): ◎ 432.872655746 => ◎ 432.872655746 (+◎ 0.000000000 0.000000000%) Some(CalculationDetail { epochs: 0, voter: 9Gko8QZBbV5SrEvHKtQHcMrGGSfgFP3KJUozEGifu25x, voter_owner: Vote111111111111111111111111111111111111111, current_effective_stake: 432868089986, total_stake: 432868089986, rent_exempt_reserve: 2282880, points: [], base_rewards: 0, commission: 10, vote_rewards: 0, stake_rewards: 0, activation_epoch: 260, deactivation_epoch: None, point_value: None, old_credits_observed: Some(4169959), new_credits_observed: Some(2349979), skipped_reasons: "ZeroCreditsAndReturnRewinded" })
```

note that no reward before&after the feature activation, new_credits_observed, and skipped_reasons.

File: ledger-tool/src/main.rs
Comment (ryoqun) [inline ledger-tool/src/main.rs:2675]: bad pasta

File: ledger-tool/src/main.rs
Comment (ryoqun) [inline ledger-tool/src/main.rs:2701]: bad pasta

**Comment by ryoqun** [comment]
> I ran `solana-ledger-tool capitalization --warp-epoch +1` against mainnet-beta. and it seems I'm getting correct inflation tracer outputs:
> 
> ```
> ## before
> 
> 5mHWstuuRk7DVPPVTZNYM43Se9czqBeZ6Woo7QrLi9CD (Stake11111111111111111111111111111111111111): ◎ 432.872655746 => ◎ 432.872655746 (+◎ 0.000000000 0.000000000%) Some(CalculationDetail { epochs: 0, voter: 9Gko8QZBbV5SrEvHKtQHcMrGGSfgFP3KJUozEGifu25x, voter_owner: Vote111111111111111111111111111111111111111, current_effective_stake: 432868089986, total_stake: 432868089986, rent_exempt_reserve: 2282880, points: [], base_rewards: 0, commission: 10, vote_rewards: 0, stake_rewards: 0, activation_epoch: 260, deactivation_epoch: None, point_value: None, old_credits_observed: Some(4169959), new_credits_observed: None, skipped_reasons: "ZeroCreditsAndReturnCurrent/ZeroPoints" })
> 
> ## after
> 
> 5mHWstuuRk7DVPPVTZNYM43Se9czqBeZ6Woo7QrLi9CD (Stake11111111111111111111111111111111111111): ◎ 432.872655746 => ◎ 432.872655746 (+◎ 0.000000000 0.000000000%) Some(CalculationDetail { epochs: 0, voter: 9Gko8QZBbV5SrEvHKtQHcMrGGSfgFP3KJUozEGifu25x, voter_owner: Vote111111111111111111111111111111111111111, current_effective_stake: 432868089986, total_stake: 432868089986, rent_exempt_reserve: 2282880, points: [], base_rewards: 0, commission: 10, vote_rewards: 0, stake_rewards: 0, activation_epoch: 260, deactivation_epoch: None, point_value: None, old_credits_observed: Some(4169959), new_credits_observed: Some(2349979), skipped_reasons: "ZeroCreditsAndReturnRewinded" })
> ```
> 
> note that no reward before&after the feature activation, new_credits_observed, and skipped_reasons.

fyi: i've rebased out-of-tree pretty hacky inflation validation subcommand of `solana-ledger-tool`.. code is here: https://github.com/ryoqun/solana/tree/ledger-tool-inflation-validation

**Comment by ryoqun** [comment]
Also, around +0.008% stakes will be started to earn staking rewards after this feature activation. this should be negligible overall

```
$ grep ReturnRewi mainnet-beta-272-new.log | awk '{print $3}' | tr -d ◎  | paste -d+ -s | bc
31806.817473008
$ ./target/release/solana -um stake-history

  Epoch       Effective Stake      Activating Stake    Deactivating Stake
    273   386164845.578965962     3751481.729363170     3375889.729314109 SOL
    272   389409318.788984954     2432030.220285488     5802706.327593587 SOL
    271   392955548.748773754     1980704.748732860     5653660.032072398 SOL
    270   393719463.247481167     2391855.557573695     3283877.238107522 SOL
    269   394411986.403078020     3051392.704833432     3872033.135666619 SOL
    268   393354958.271813154     2592345.372506808     1663351.552270411 SOL
    267   395165510.789034367     2032117.252118937     3969902.194922559 SOL
    266   395559881.128125072     4024142.877033422     4545366.195989004 SOL
    265   394969262.506878734     3243355.053764615     2780222.688156187 SOL
    264   396038677.696802497     1624993.008589262     2822182.589519491 SOL
$ pry
[1] pry(main)> "%0.5f%%" % (100 * 31806.817473008 / 386164845.578965962) 
=> "0.00824%"

```

**Comment by ryoqun** [comment]
(status: this pr is almost ready for review. it went through rather rigorous self-review including full re-audit of staking code)

File: programs/stake/src/stake_state.rs
Comment (ryoqun) [inline programs/stake/src/stake_state.rs:3529]: seems outdated comment.

File: programs/stake/src/stake_state.rs
Comment (ryoqun) [inline programs/stake/src/stake_state.rs:322]: a bit contrived flag thing. but i rather wanted some encapsulation for maintainability (place new code touched at lowest-level for all codepath even including currently-hypothetical code, which isn't via this `fn calculate_stake_rewards`) and separation-of-concerns (no visible muddy credits handling with feature activation here and put it into `calculate_stake_point_and_credits`)

**Comment by ryoqun** [comment]
@CriesofCarrots this is finally ready for review. :)

File: programs/stake/src/stake_state.rs
Comment (CriesofCarrots) [inline programs/stake/src/stake_state.rs:221]: Is it time to make this return type a struct?

File: programs/stake/src/stake_state.rs
Comment (CriesofCarrots) [inline programs/stake/src/stake_state.rs:320]: Thoughts of making this return type an `Option<struct>` as well, while you're in here? (I have to keep scrolling all over the place to remember what the values are)

**Review by CriesofCarrots — DISMISSED** [review]
Thank you, @ryoqun , this looks great! Just wondering if it's time for some helper return structs to keep values clear, but I'm okay to ship as-is, if you want to keep the changeset smaller.

File: programs/stake/src/stake_state.rs
Comment (ryoqun) [reply programs/stake/src/stake_state.rs:221]: yeah, i tempted doing the cleanup. but, the tuple is still acceptable, imo...

File: programs/stake/src/stake_state.rs
Comment (ryoqun) [reply programs/stake/src/stake_state.rs:320]: sorry but I'll pass on.

Comment (CriesofCarrots) [reply programs/stake/src/stake_state.rs:320]: 😢

File: runtime/src/bank.rs
Comment (t-nelson) [inline runtime/src/bank.rs:2908]:
```suggestion
    pub fn unfreeze_for_ledger_tool(&self) {
```

File: programs/stake/src/stake_state.rs
Comment (t-nelson) [inline programs/stake/src/stake_state.rs:251]: :+1:

**Review by t-nelson — DISMISSED** [review]
lgtm!

**Comment by ryoqun** [comment]
status: this pr is pended because of questioned risk/benefits for changing on this (https://github.com/solana-labs/solana/issues/22512#issuecomment-1029291043 ; as a stake-related engineer, i think this pr is safe and might be beneficial but not so strongly opinionated to ship this) and i requested audits to @bennofs (indicated he can find something interesting in this change)

**Review by bennofs — APPROVED** [review]
Thanks for waiting for me, I reviewed the changes and it looks good to me. My reasoning for why this is safe:

* the rewind code is only triggered when there is a vote account with lower credits than observed in a rooted bank. This means that there is no way that any vote state with higher credits observed is obtainable.
* the codepath that calls the rewind code reloads the stake account from the bank, so even if there were inconsistencies in the stakes cache these wouldn't matter here
* rewinding only happens at the end of the epoch, and no rewards are paid for ending epoch. So the outcome is the same as you would get by just unstaking and restaking

I also did a quick search for credits_observed in the code base just to make sure that we didn't forget about something that uses that field and would break, but only found the stake program and some cli tools/tests so that's good as well.

**Comment by CriesofCarrots** [comment]
@ryoqun , what are you thinking about this one?

**Comment by ryoqun** [comment]
> @ryoqun , what are you thinking about this one?

@CriesofCarrots  thanks for caring this pr with love!

admittedly, this pr was stale... I just has been focusing on other stuff. Specifically, I'm wondering how should we weigh in the relieve efforts related to competencies of users. There were some proceedings of stalemate for similar topics:

- base58c ref: https://github.com/solana-labs/solana/issues/6970
- ATA's closeNested: https://github.com/solana-labs/solana-program-library/pull/2889

I still think that this pr is safe (now that it even got an approval from allies/neodyme!) to ship but it needs somewhat rigorious testing with testnet/devnet to actually enable this feature ultimately in mainnet-beta and monitor several epochs there. (it's not definitely just switch-and-forget thing..) So, I was a bit reluctant to invest time to ship this eagerly recently.

**Comment by CriesofCarrots** [comment]
I still think we should make this fix, as current behavior is objectively broken. Assuming I can push to this branch, I will rebase here.
@ryoqun , can you please give me an update on any testing you've done, or think is required? Thank you.

**Comment by t-nelson** [comment]
oh i didn't realize we never landed this :confused:

File: runtime/src/bank.rs
Comment (CriesofCarrots) [reply runtime/src/bank.rs]: Updated

**Comment by ryoqun** [comment]
> I still think we should make this fix, as current behavior is objectively broken. Assuming I can push to this branch, I will rebase here.

@CriesofCarrots thanks for taking over and rebasing this. it's embarrassing for me not to be able to work on this for such long time. i have been focusing on the prolonged cluster-wide instability with primary focus of assisting affected ecosystem partners. maybe, it's time for me to take a breather and address not-highest-priority tasks...

as @mvines reacted with `+1`, i guess team's consensus here is that this fix outweigh [this concern](https://github.com/solana-labs/solana/issues/22512#issuecomment-1029291043) at this moment. :)

> @ryoqun, can you please give me an update on any testing you've done, or think is required? Thank you.

as said above, i haven't spent time on this regarding testing... And, I think it's required, considering the touched code. Once this lands (with guarantee of shipping eventually), i'll report back here with results with procedures (i.e. csv, its `solana-ledger-tool` one-liner, a short commentary from me). 1.11.x is targeted without backporting, maybe? (I'll do that testing with tip of `master` after merge, then.)

Also, I'll monitor few epochs before and after around devnet/testnet feature activation.

lastly, sorry for not properly communicating about this pr's stalemate.

File: runtime/src/bank.rs
Comment (ryoqun) [inline runtime/src/bank.rs:3328]: :+1:

**Review by t-nelson — APPROVED** [review]

**Comment by ryoqun** [comment]
> > Once this lands (with guarantee of shipping eventually), i'll report back here with results with procedures (i.e. csv, its `solana-ledger-tool` one-liner, a short commentary from me). 1.11.x is targeted without backporting, maybe? (I'll do that testing with tip of `master` after merge, then.)
> 
> Thank you, that sounds perfect. Yes, I'm planning to let this ride in on v1.11 without backporting.

fyi, I'm still not forgetting this pr and this live-cluster verification chore. :) I'm just waiting v1.11 upgrade for devnet and testnet.

**Comment by ryoqun** [comment]
> > > Once this lands (with guarantee of shipping eventually), i'll report back here with results with procedures (i.e. csv, its `solana-ledger-tool` one-liner, a short commentary from me). 1.11.x is targeted without backporting, maybe? (I'll do that testing with tip of `master` after merge, then.)
> > 
> > 
> > Thank you, that sounds perfect. Yes, I'm planning to let this ride in on v1.11 without backporting.
> 
> fyi, I'm still not forgetting this pr and this live-cluster verification chore. :) I'm just waiting v1.11 upgrade for devnet and testnet.

@CriesofCarrots this is done with no concerns found. I've found no not-explainable wierdness  after carefully investigating actual inflation distributions after the feature activation on testnet. 

(for the very curious stranger who is eager to redo the verification step, this is done with rebased out-of-tree ugly inflation validation code, which I'm carrying in my bag for years: (found here: https://github.com/ryoqun/solana/commit/53436442d6b3acb17ea3f78cdb1f301d4ebac5e8) and the dataset used was here https://drive.google.com/drive/folders/13DE2dvdfhfhN5VWNIDaJQKGgOLmWr4k9?usp=sharing)

So, devnet and mainnet-beta is remaining..

**Comment by ryoqun** [comment]
this feature finally was activated. in short, it seems working as expected:

```
$ grep ReturnRewi ./mainnet-beta-after-rewind-with-rewind.log | grep -o current_effective_stake.*$ | awk '{print $2}' | tr -d , | paste -d+ -s | bc
22500153335748
$ solana -um feature status --display-all BUS12ciZ5gCoFafUHWW8qaFMMtwFQGVxjsDheWLdqBE2
Feature                                      | Status                  | Activation Slot | Description
BUS12ciZ5gCoFafUHWW8qaFMMtwFQGVxjsDheWLdqBE2 | active since epoch 464  | 200448008       | Auto rewind stake's credits_observed if (accidental) vote recreation is detected #22546
$ solana -um stake-history
  Epoch       Effective Stake      Activating Stake    Deactivating Stake
    464   388620608.562748790     1797516.702643046     1265885.799712572 SOL
    463   388283431.704100668     1343627.726265571     1121176.049718931 SOL
    462   387366397.995904684     2710135.622854811     1907508.178961365 SOL
    461   388222876.775501192     2401705.256340335     3372082.311824944 SOL
    460   389072050.175475955     2000915.199247054     2964338.113380553 SOL
$ irb
irb(main):002:0> "%0.5f%%" % (100 * 22_500_153_335_748.0 / 388_620_608_562_748_790) 
=> "0.00579%"
```

(this is redo of https://github.com/solana-labs/solana/pull/22546#issuecomment-1025152266)
 
> Also, around +0.007% (currently stalled due to the accidental vote account recreation) stakes will be started to earn staking rewards after this feature activation. this should be negligible overall

1.5 years has passed. +0.006% started to earn rewards; seems the total staked amount is mostly unchanged and the stuck stake amount reduced because some of newly recreated vote accounts surpassed credits of deleted vote accounts.

**Comment by ryoqun** [comment]
> > I ran `solana-ledger-tool capitalization --warp-epoch +1` against mainnet-beta. and it seems I'm getting correct inflation tracer outputs for sampled stake account `5mHWstuuRk7DVPPVTZNYM43Se9czqBeZ6Woo7QrLi9CD`:

> the stuck stake amount reduced because some of newly recreated vote accounts surpassed credits of deleted vote accounts.

as this sampled vote account had already been earning staking rewards prior to this feature activation, i've chosen different sampled stake account: `HVGsbFNgwaniyJNQY9bP4LxqpyR4V2Z9yqDdVf71ToDL`:

```
$ solana -um stake-account --with-rewards --num-rewards-epochs 5 HVGsbFNgwaniyJNQY9bP4LxqpyR4V2Z9yqDdVf71ToDL --verbose
RPC URL: https://api.mainnet-beta.solana.com
Default Signer Path: /home/ryoqun/.config/solana/id.json
Commitment: confirmed
Balance: 344.289928228 SOL
Rent Exempt Reserve: 0.00228288 SOL
Delegated Stake: 344.278513828 SOL
Active Stake: 344.278513828 SOL
Delegated Vote Account Address: 7gfSX88DYoScBCaFqNxbsgFeZom8Qotnf747iJLhhJC6
Stake Authority: FNsitHktYun9mksohjdDgE5YGaGByDyJJAeFEeJHSHTv
Withdraw Authority: EhYXq3ANp5nAerUpbSgd7VK2RRcxK1zNuSQ755G5Mtxx
Lockup Timestamp: 2027-02-07T00:00:00Z
Lockup Custodian: 3XdBZcURF5nKg3oTZAcfQZg8XEc5eKsx6vK8r3BdGGxg
Epoch Rewards:
  Epoch   Reward Slot  Time                        Amount              New Balance         Percent Change             APR  Commission
  465     201312012    2023-06-24 07:00:19 +09:00  ◎0.145596747        ◎344.289928228           0.042307%           6.75%         10%
  464     200880004    2023-06-22 00:07:28 +09:00  ◎0.145802933        ◎344.144331481           0.042385%           6.84%         10%
Credits Observed: 55886414

$ solana -um vote-account 7gfSX88DYoScBCaFqNxbsgFeZom8Qotnf747iJLhhJC6
Account Balance: 27.191143833 SOL
Validator Identity: 6GRLDLiAtx8ZjYgQgPo7UsYeJ9g1pLX5j3HK97tFmtXb
Vote Authority: {466: "6GRLDLiAtx8ZjYgQgPo7UsYeJ9g1pLX5j3HK97tFmtXb"}
Withdraw Authority: DtutfyBQVgxn2AHSq74E5TFfB2fsnP4peRYtp3uLK2tz
Credits: 55959049
Commission: 10%
Root Slot: 201388426
Recent Timestamp: 2023-06-24T07:38:21Z from slot 201388469
Recent Votes (using 31/31 entries):
- slot: 201388469 (confirmation count: 1)
- slot: 201388468 (confirmation count: 2)
- slot: 201388467 (confirmation count: 3)
- slot: 201388466 (confirmation count: 4)
- slot: 201388465 (confirmation count: 5)
- slot: 201388464 (confirmation count: 6)
- slot: 201388457 (confirmation count: 7)
- slot: 201388450 (confirmation count: 8)
- slot: 201388449 (confirmation count: 9)
- slot: 201388448 (confirmation count: 10)
- slot: 201388447 (confirmation count: 11)
- slot: 201388446 (confirmation count: 12)
- slot: 201388445 (confirmation count: 13)
- slot: 201388444 (confirmation count: 14)
- slot: 201388443 (confirmation count: 15)
- slot: 201388442 (confirmation count: 16)
- slot: 201388441 (confirmation count: 17)
- slot: 201388440 (confirmation count: 18)
- slot: 201388439 (confirmation count: 19)
- slot: 201388438 (confirmation count: 20)
- slot: 201388437 (confirmation count: 21)
- slot: 201388436 (confirmation count: 22)
- slot: 201388435 (confirmation count: 23)
- slot: 201388434 (confirmation count: 24)
- slot: 201388433 (confirmation count: 25)
- slot: 201388432 (confirmation count: 26)
- slot: 201388431 (confirmation count: 27)
- slot: 201388430 (confirmation count: 28)
- slot: 201388429 (confirmation count: 29)
- slot: 201388428 (confirmation count: 30)
- slot: 201388427 (confirmation count: 31)
- ... (truncated 55959049 rooted votes, which have been credited)
Recent Epoch Voting History (using 64/64 entries):
* missed credits include slots unavailable to vote on due to delinquent leaders
- epoch: 466
  credits range: (55886414..55959049]
  credits/slots: 72635/432000
- epoch: 465
  credits range: (55474859..55886414]
  credits/slots: 411555/432000
- epoch: 464
  credits range: (55059628..55474859]
  credits/slots: 415231/432000
- epoch: 463
  credits range: (54642620..55059628]
  credits/slots: 417008/432000
- epoch: 462
  credits range: (54227245..54642620]
  credits/slots: 415375/432000

```

you can see this account started to earn reward only after epoch 464 (this feature got activated) and onwards.
the rewind rewrite was done at the _start_ of epoch 464 instantly after feature activation. so, it's expected that any rewound stake accounts earn rewards at the _end_ of epoch 464.

After that, these stake accounts are just normal stake accounts. so, I've seen no rewound stakes accounts at epoch 465.

**Comment by ryoqun** [comment]
also, I've checked all stuck stake accounts are normalized to same credits by vote accounts (there were 4 bad vote accounts at the time of feature activation):

```
$ grep ReturnRewi ./mainnet-beta-after-rewind-with-rewind.log | grep -E -o 'voter:[^,]*|new_credits_observed[^,]*' | sort | uniq -c | sort -n
     16 new_credits_observed: Some(0)
     16 voter: 7X5QAMP54upioBdu75QZkz8hpYcwz7UmzAHGB2STq9da
     77 new_credits_observed: Some(55048232)
     77 voter: 7gfSX88DYoScBCaFqNxbsgFeZom8Qotnf747iJLhhJC6
    213 new_credits_observed: Some(38209189)
    213 voter: 3Xn3K5zeTzs4sURj1PCRaF6rvFg2hPgT49B4SiM998f7
    483 new_credits_observed: Some(345583)
    483 voter: 47e2BfrzbWdDieFZA7fc9fX1f73WTtkok2fQxSb8GueY

```

**Pull request merged**
